### PR TITLE
update go-multistream and use negotiateLazy 

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

File: p2p/host/basic/basic_host.go [hunk 1]
```diff
@@ -104,6 +104,7 @@ func New(net inet.Network, opts ...interface{}) *BasicHost {
 
 // newConnHandler is the remote-opened conn handler for inet.Network
 func (h *BasicHost) newConnHandler(c inet.Conn) {
+	h.Peerstore().SetProtocols(c.RemotePeer())
```
Comment (whyrusleeping) [inline p2p/host/basic/basic_host.go]: we have issues around remembering protocols across connection resets. For now we're just clearing them

File: p2p/host/basic/basic_host_test.go [hunk 2]
```diff
@@ -98,7 +98,7 @@ func TestHostProtoPreference(t *testing.T) {
 	protoNew := protocol.ID("/testing/1.1.0")
 	protoMinor := protocol.ID("/testing/1.2.0")
 
-	connectedOn := make(chan protocol.ID, 16)
+	connectedOn := make(chan protocol.ID)
```
Comment (whyrusleeping) [inline p2p/host/basic/basic_host_test.go]: having the buffer here was allowing some 'slack' in the test which hid failures

File: p2p/host/basic/basic_host.go (hunk 1 above)
Comment (Kubuxu) [inline p2p/host/basic/basic_host.go]: Do you want to put it in comment in there? It is not really clear from the code.

**Comment by Kubuxu** [comment]
Also travis is failing, is it random or connected?

**Pull request merged**